services/provider/server/server.go

```diff
@@ -5,43 +5,202 @@ import (
+	}
+
+	uid, err := s.createStorageConsumer(req.ConsumerName, token, capacity)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to create storageConsumer resource %q: %v", req.ConsumerName, err)
+	}
+
+	consumer := consumer.NewConsumer(req.ConsumerName, uid, token)
+	err = s.authManager.consumerStore.Save(consumer)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to update consumer store: %v", err)
+	}
+
+	encryptedUID, err := s.authManager.GetEncryptedID(uid)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to encrypt consumer resource ID: %v", err)
+	}
+
+	// TODO: send correct granted capacity
+	return &pb.OnBoardConsumerResponse{StorageConsumerUUID: encryptedUID, GrantedCapacity: req.Capacity}, nil
 }
 
 // GetStorageConfig RPC call to onboard a new OCS consumer cluster.
-func (c *ocsProviderServer) GetStorageConfig(ctx context.Context, req *pb.StorageConfigRequest) (*pb.StorageConfigResponse, error) {
-	return &pb.StorageConfigResponse{}, nil
+func (s *OCSProviderServer) GetStorageConfig(ctx context.Context, req *pb.StorageConfigRequest) (*pb.StorageConfigResponse, error) {
+	consumerResource, err := s.getStorageConsumer(req.StorageConsumerUUID)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to validate storage consumer: %v", err)
+	}
+
+	// Verify Status
+	switch consumerResource.Status.State {
+	case ocsv1alpha1.StorageConsumerStateFailed:
+		// TODO: get correct error message from the storageConsumer status
+		return nil, status.Errorf(codes.Internal, "storageConsumer status failed")
+	case ocsv1alpha1.StorageConsumerStateConfiguring:
+		return nil, status.Errorf(codes.Unavailable, "waiting for the rook resources to be provisioned")
+	case ocsv1alpha1.StorageConsumerStateDeleting:
+		return nil, status.Errorf(codes.Internal, "storageConsumer is already in deleting phase")
+	case ocsv1alpha1.StorageConsumerStateReady:
+		// TODO: Return json connection details
+		return &pb.StorageConfigResponse{}, nil
+	}
+
+	return nil, status.Errorf(codes.Unavailable, "storage consumer status is not set")
 }
 
 // OffBoardConsumer RPC call to delete the StorageConsumer CR
-func (c *ocsProviderServer) OffBoardConsumer(ctx context.Context, req *pb.OffBoardConsumerRequest) (*pb.OffBoardConsumerResponse, error) {
+func (s *OCSProviderServer) OffBoardConsumer(ctx context.Context, req *pb.OffBoardConsumerRequest) (*pb.OffBoardConsumerResponse, error) {
+	consumerResource, err := s.getStorageConsumer(req.StorageConsumerUUID)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to validate storage consumer: %v", err)
+	}
+
+	// Delete storageConsumer resource no matter the current state of the resource
+	err = s.consumerAPI.DeleteStorageConsumer(consumerResource)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to delete storage consumer resource: %v", err)
+	}
+
```

Comment:
Where are we going to protect again deletion of StorageConsumers that are still consuming storage (have claims on storage)?

Reply:
Should we look for the available PVCs (consuming ocs storage) in the consumer and send the offboarding request only if not exists?

Reply:
No, the source of truth should always be the provider. 
@leseb is there anything in rook-ceph we can query to get that information (for both RBD and CephFS)

Reply:
Consumer Storage Cluster is deleted ->  Consumer storage cluster deletes the resources its using (RBD/Ceph FS) -> Only after successful deletion of those resources the consumer OCS cluster will call the Offboarding API to delete the StoageConsumer CR . 

This ^^ is one possible approach that we discussed today.

Reply:
@sp98 The API server is not an internal component, it is a public endpoint, and as such cannot assume the correct "order of operations" is enforced from the outside. 

It is true that for the "happy path" the strategy you describe can work, but what about bugs? what about attackers? 
The risk here is loss of customer data and as such we must protect against such situations.

Reply:
From the Rook perspective, once the CephBlockPool and SubVolumeGroup are in deleting phase, the finalizers will only be released once no RBD image and no SubVolume are present anymore. Until this is the case, the deletion of these CRDs will be blocked. Today, the failure is not reflected in the CR status, but it's quite easy to do. I'll add this.

Reply:
@leseb 
I think the CR status is too late as the consumer will get an ack on the delete (as we agreed that it is an async process). We need to fail the delete request before we even change the storageconsumer CR